### DISCUSSION TITLE
Alternative BFT algorithms

### DISCUSSION THREAD

**Discussion by justinmoon** [fedimint/fedimint#1791] [General]
...

**Comment by elsirion** [comment]
My favorite so far once we find the time (ofc given there are no weaknesses discovered): [Bolt-Dumbo Transformer: Asynchronous Consensus As Fast As the Pipelined BFT](https://arxiv.org/abs/2103.09425)

**Comment by joschisan** [comment]
In the HBBFT paper, they optimize for high throughput and latency while Fediment requires low latency and does not require high throughput if I understand the intended applications correctly. The authors assume the system to be under load and the transaction queues to be full most of the time. Many of the cryptographic primitives are used to increase throughput not to achieve asynchronicity and may need large batch sizes to amortize. However, in the case of fedimint, I would expect most queues to be empty most of the time. Therefore I was wondering if the following two elements of the protocol actually make sense in our case.


The first example is the reliable broadcast protocol. The authors write

>While Bracha’s [13] classic reliable broadcast protocol requires O(N2|v|) bits of total communication in order to broadcast a mes- sage of size |v|, Cachin and Tessaro [18] observed that erasure coding can reduce this cost to merely O(N|v|+λN2 logN), even in the worst case. This is a significant improvement for large messages (i.e., when |v| ≫ λ N log N ), which, (looking back to Section 4.3) guides our choice of batch size.

Since I assume most batches to be empty or contain a single transaction I am wondering if erasure coding is actually an overall improvement.

The second example is the use of threshold encryption. It is my understanding that its only use is censorship resistance and is necessary as HBBFT chooses transactions not from the beginning of the queue but at random to achieve higher throughput as different nodes likely propose different transactions. This assumes sufficiently full queues. However, In our case, we could simply choose the first n elements as our queues are mostly empty or contain only a few elements anyways. The authors write

>The cost of ACS depends directly on size of the transaction sets proposed by each node. Since the output vector contains at least N − f such sets, we can therefore improve the overall efficiency by ensuring that nodes propose mostly disjoint sets of transactions, thus committing more distinct transactions in one batch for the same cost. Therefore instead of simply choosing the first element(s) from its buffer (as in CKPS01 [15]), each node in our protocol proposes a randomly chosen sample, such that each transaction is, on average, proposed by only one node.
However, implemented naïvely, this optimization would compromise censorship resilience, since the ACS primitive allows the adversary to choose which nodes’ proposals are ultimately included. The adversary could selectively censor a transaction excluding whichever node(s) propose it. We avoid this pitfall by using threshold encryption, which prevents the adversary from learning which transactions are proposed by which nodes, until after agreement is already reached.

Therefore I am wondering if threshold encryption is necessary as clients are expected to send a transaction to as many nodes as necessary anyways. It seems to me that this alone can prevent censorship if the nodes choose transactions from the front of the queue.

It seems to me that we could forego erasure coding and threshold encryption to reduce computational load on the nodes and code complexity. As we do not require high throughput and do not have sufficiently full queues this may fit our use case better.

**Reply by elsirion** [reply]
First of all thank you for looking into the issue so thoroughly and sorry for the late reply.

> Since I assume most batches to be empty or contain a single transaction I am wondering if erasure coding is actually an overall improvement.

Quite possibly not, but having an implementation available was a big part of choosing a BFT algo. I also think while most basic e-cash federations will need little throughput others could experience higher load (e.g. if someone implements a federated exchange).

> Therefore I am wondering if threshold encryption is necessary as clients are expected to send a transaction to as many nodes as necessary anyways. It seems to me that this alone can prevent censorship if the nodes choose transactions from the front of the queue.

That sounds reasonable, when I did performance testing that was always under high load and while I experienced some performance degradation from threshold encryption it was minor compared to e-cash issuance. Now that we optimize for latency it would make sense to revisit these points.

I think a prerequisite for any BFT improvements is making the BFT algo plug'n'play, meaning creating a `DynBftAlgo` type that can wrap any BFT impl that conforms to a common interface. Then we could keep HBBFT proper for more conservative operators but also work on a Fedimint-HBBFT-lite impl that has better performance. This would also allow us to experiment with synchronous and hybrid BFT algos eventually.

The details might be a bit more involved since depending on the BFT algo there would be different config required, but that shouldn't be any harder to implement than the module system we already have. Do you see any further pitfalls @jkitman and @dpc?

**Reply by jkitman** [reply]
>  I am wondering if threshold encryption is necessary as clients are expected to send a transaction to as many nodes as necessary anyways. It seems to me that this alone can prevent censorship

That's a good point.

We don't rely on any details of the consensus algorithm so I think it wouldn't be too hard to create a common interface for it.

**Comment by joschisan** [comment]
I have just found another small tweak that could possibly improve your latency. In hbbft the authors write in Algorithm ACS:

> upon delivery of value 1 from at least N − f instances of BA, provide input 0 to each instance of BA that has not yet been
provided input.

Waiting for the completion of a single BA with 1 is sufficient to ensure progress at every round but cannot guarantee censorship resistance if done naively. However, it appears to me that waiting for only f + 1 instances of BA to complete with 1 should be sufficient to prevent censorship from f faulty nodes conspiring with the adversary in control of the network. I assume that hbbft waits for N - f instances to complete with 1 as this is the maximum number such that the algorithm can still make progress with f byzantine nodes and this probably optimizes throughput as hbbft is designed so that nodes propose mostly distinct sets which it seems to me is not the case with a lower throughput application like e-cash. 

In the paper Dumbo: Faster Asynchronous BFT Protocols the authors identify the ABA instances as the major bottleneck for time per round (Figure 2) when compared to the reliable broadcast for 16 nodes and up. One main reason for this is that the ABA instances do not proceed independently as mentioned previously. In the worst case of f faulty nodes, every correct broadcast by a correct node and corresponding ABA instance has to complete before the correct nodes make any inputs to the other ABA instances. Hence we are waiting for the slowest broadcast and ABA to complete. In that case, there may be a decrease in round time if we only have to wait for f + 1, so roughly the faster half, or just 1, so only the fastest. One would expect the improvement to increase with the deviation between broadcast / ABA completion times and the number of faulty nodes.
